lib/trino-collect/src/main/java/io/trino/collect/cache/EvictableCache.java

```diff
@@ -145,31 +148,31 @@ public ImmutableMap<K, V> getAll(Iterable<? extends K> keys)
     {
         List<Token<K>> newTokens = new ArrayList<>();
         try {
-            BiMap<K, Token<K>> keyToToken = HashBiMap.create();
+            Map<K, V> result = new LinkedHashMap<>();
             for (K key : keys) {
+                if (result.containsKey(key)) {
+                    continue;
+                }
                 // This is not bulk, but is fast local operation
                 Token<K> newToken = new Token<>(key);
                 Token<K> token = tokens.computeIfAbsent(key, ignored -> newToken);
-                keyToToken.put(key, token);
-                if (token == newToken) {
-                    newTokens.add(newToken);
+                if (token != newToken) {
+                    // Token exists but a data may not exist (e.g. due to concurrent eviction)
+                    V value = dataCache.getIfPresent(token);
+                    if (value != null) {
+                        result.put(key, value);
+                        continue;
+                    }
                 }
+                newTokens.add(newToken);
             }
 
-            Map<Token<K>, V> values = dataCache.getAll(keyToToken.values());
-
-            BiMap<Token<K>, K> tokenToKey = keyToToken.inverse();
-            ImmutableMap.Builder<K, V> result = ImmutableMap.builder();
+            Map<Token<K>, V> values = dataCache.getAll(newTokens);
```

Comment:
Here we are loading for `newToken`s even if previous `token` was already present in `tokens` map. 
Do I understand correctly that if there was a case that `token` was already present in `tokens` map and value mapped to it was `null` (we did not get into `continue` in line 164) we assume that removal listener was just not called yet?
And if we put mapping for `newToken` it will start being effective soon (when  removal listener is actually called)?

Reply:
> Do I understand correctly that if there was a case that token was already present in tokens map and value mapped to it was null (we did not get into continue in line 164) we assume that removal listener was just not called yet?

yes

> And if we put mapping for newToken it will start being effective soon (when removal listener is actually called)?

no. it there is an existing `token`, the `newToken` isn't registered in `tokens`, so no other call to the cache will use it.

Reply:
> Can there be a "lag" between return from `getAll` and actually observing cached value on next calls to `get*`

no, the entries are inserted into the cache before getAll returns, no async thread

Reply:
> no. it there is an existing token, the newToken isn't registered in tokens, so no other call to the cache will use it.

As we discussed we can do better (or worse depending on point of view as the "better" is more complex) and swap `old-token` with `new-token` in `tokens` map in case when no value was assigned to old-token, and after we did `dataCache.getAll(newTokens);`. 
Similar stuff in `get()` I suppose

Reply:
i added invalidation of those dandling new tokens
didn't want to replace old tokens since it can be some other concurrent load happening

added some comment to indicate this isn't only about concurrent eviction, but also about concurrent non-bulk load